keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+kind: BannedWords
+metadata:
+  name: "banned-words.example.com"
+spec:
+  bannedWords:
+  - glitter
+  - rainbow
+```
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "policy1.example.com"
+spec:
+  match:
+    rules:
+    - apiGroups:   ["*"]
+      apiVersions: ["*"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+      scope:       "*"
+  validations:
+  - rule: "!object.name in config.bannedWords"
+```
+
+```yaml
+apiVersion: admissionregistration.k8s.io/v1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "policy2.example.com"
+spec:
+  match:
+    rules:
+    - apiGroups:   [""]
+      apiVersions: ["v1"]
+      operations:  ["CREATE", "UPDATE"]
+      resources:   ["pods"]
+      scope:       "*"
+  validations:
+  - rule: "!object.spec.containers.any(c, c.name in config.bannedNames)"
+  - rule: "!object.spec.initContainers.any(c, c.name in config.bannedNames)"
+```
+
+
+Similar Use Case: A cluster administrator wishes to use a single policy
+configuration to manage a network policy that must be enforced across multiple
+Kubernetes kinds that contain relevant networking fields. It is possible to
+implement by having multiple `ValidatingAdmissionPolicy` resources that all
+reference the same `spec.config` CRD but that each enforce the policy for a
+different Kubernetes network kind.
+
+#### Use Case: Principle of least privilege policy
+
+A cluster administrator would like disallow the use of a list of reserved labels
+by default, but allow use of the labels in specific namespaces so long as the
+label values are valid. 
+
+- Define the following policy configurations:
+  - policy configuration that matches all namespaces with the
+    "may-use-reserved-labels" label and checks that the values are valid
+  - "default" policy configuration that matches all namespaces without the
+    "may-use-reserved-labels" label and disallows any reserved labels.
+  - An additional policy can be used validate that only authorized roles set the
+    "may-use-reserved-labels".
+
+See "Default configuration
+
+#### Use Case: Validating native type with new field (version skew case)
+
+Policy author wants to write a policy that validates a property of all
+containers in a pod, including ephemeralContainers for versions of
+Kubernetes where ephemeralContainers are available.
+
+```yaml
+  validations:
+    - expression: "object.spec.containers.all(c, c.name.startsWith('xyz-'))"
+    - expression: "!has(object.spec.initContainers) || object.spec.initContainers.all(c, c.name.startsWith('xyz-'))"
+    - expression: "!has(object.spec.ephemeralContainers) || object.spec.ephemeralContainers.all(c, c.name.startsWith('xyz-'))"
+```
+
+This does not work if type checking is strict. The ephemeralContainers field
+will be reported as unrecognized.
+
+This becomes far more compact if we offer scoping:
+
+```yaml
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
```

Comment:
I don't think I find this easier to read.

Reply:
Added a note to the UNRESOLVED.